### PR TITLE
Local cam

### PR DESCRIPTION
**Pull request by stefan-sherwood**
image

image

### CODE REVIEW

**Comment by stefan-sherwood** [comment]
Notes:
- Camera doesn't disconnect stream when tab is inactive (TODO)
- Close camera button (upper-right corner) is too small and not obvious enough
- Edit box and dropdown are different sizes and don't line up properly. Not obvious to me how to fix but haven't spent much time on it

Some possible future features:
- Multiple cams at the same time
- Video popup (e.g. reverse cam) over other tabs
- Trigger popup with shortcut/keystroke
- Auto connect on startup
- Still snapshot
- Record video

File: src/app/tabs/camera.cpp [hunk 1]
```diff
@@ -1,135 +1,313 @@
-#include <QLineEdit>
```
Comment (stefan-sherwood) [inline src/app/tabs/camera.cpp:1]: I don't know why it thinks the entire file changed (it didn't)

File: src/app/tabs/camera.cpp [hunk 2]
```diff
@@ -1,135 +1,313 @@
...
+    this->player->setMedia(QUrl(pipeline));
+    this->player->play();
+}
+
+void CameraTab::connect_local_stream()
+{
+    if (this->local_cam != nullptr)
+        delete this->local_cam;
+
+    if (!this->local_cam_available(this->local_device)) {
```
Comment (stefan-sherwood) [inline src/app/tabs/camera.cpp:264]: If you give QCamera a bogus device path it just connects to the default camera with no complaint. This code is an attempt to make behavior more predictable, particularly when connecting to a camera that was disconnected without the camera list dropdown being refreshed.

**Comment by stefan-sherwood** [comment]
~~Camera doesn't disconnect stream when different tab is active (TODO)~~

It turns out that the camera stream uses virtually no resources when the video picture isn't visible.

**Comment by rsjudka** [comment]
> Close camera button (upper-right corner) is too small and not obvious enough

what scale factor are you using? Seems fine when using 1x scale but i would agree anything below that it looks a bit small

**Comment by rsjudka** [comment]
I think it would be best to avoid using dropdowns... if you see for example the "color picker" would have made sense to use a dropdown, but its nearly impossible to style and doesn't work when using the "mocked" brightness module (aka dropdown opacity doesn't update with window)

**Comment by rsjudka** [comment]
also not sure im a fan of the radio button to select between network and local... my guess would be a user would either be using one type or the other, but not both

UI should be as minimalist as possible

**Comment by stefan-sherwood** [comment]
> I think it would be best to avoid using dropdowns... if you see for example the "color picker" would have made sense to use a dropdown, but its nearly impossible to style and doesn't work when using the "mocked" brightness module (aka dropdown opacity doesn't update with window)

It would be really unusual that people would have more than one item in the dropdown, at least for now. Are you suggesting I just dynamically create entries for the different options? I can certainly do something like that.

**Comment by stefan-sherwood** [comment]
FWIW, I don't like my own UI. I was primarily doing what was expedient so that others could test its viability.

**Comment by stefan-sherwood** [comment]
> > Close camera button (upper-right corner) is too small and not obvious enough
> 
> what scale factor are you using? Seems fine when using 1x scale but i would agree anything below that it looks a bit small

I didn't change the scale factor so it's 1x. On an 8-inch screen, fat fingers can't easily hit it. It should be easy to close, even when driving.

**Comment by stefan-sherwood** [comment]
Also note that the X for closing the app is much bigger than the X for closing the tab.

**Comment by stefan-sherwood** [comment]
image

1x scale with finger for reference. IA running on 8-inch screen.

**Comment by rsjudka** [comment]
> It would be really unusual that people would have more than one item in the dropdown, at least for now. Are you suggesting I just dynamically create entries for the different options? I can certainly do something like that.

Can you explain a bit more what you mean? I was just saying it would be better to use the "spinwheel" type widget I used for thing like the color picker

**Comment by rsjudka** [comment]
> FWIW, I don't like my own UI. I was primarily doing what was expedient so that others could test its viability.

hmm this might be a good usecase for something like a "staging" repo. There we can basically add features as they get created, and then work on improvements and UI related stuff when going from staging -> develop

**Comment by rsjudka** [comment]
> 1x scale with finger for reference. IA running on 8-inch screen.

Ah yep that's small haha I guess I should have realized screen ratio != physical size

It'll be hard to try and balance the size of the button while trying to maximize the camera screen size

I'm thinking maybe we do something where if you tap the frame an x will pop-up or something? Hovering over it

**Comment by stefan-sherwood** [comment]
> > FWIW, I don't like my own UI. I was primarily doing what was expedient so that others could test its viability.
> 
> hmm this might be a good usecase for something like a "staging" repo. There we can basically add features as they get created, and then work on improvements and UI related stuff when going from staging -> develop

I thought that's what develop was for

**Comment by rsjudka** [comment]
> I thought that's what develop was for

Yea, but we I still want develop to be "clean" - if that makes sense

Basically I think develop should always be ready to merge to master, meaning the code is clean and efficient, UI is well thought, etc

Having an additional branch would allow us to get more usage of new features that still need some "non critical" pieces finished (like the UI in your case)

**Comment by stefan-sherwood** [comment]
> > I thought that's what develop was for
> 
> Yea, but we I still want develop to be "clean" - if that makes sense
> 
> Basically I think develop should always be ready to merge to master, meaning the code is clean and efficient, UI is well thought, etc
> 
> Having an additional branch would allow us to get more usage of new features that still need some "non critical" pieces finished (like the UI in your case)

That makes perfect sense. Everywhere I've worked, what you describe as develop has been the staging branch but I don't care what it's called as long as it clear to people what's what.

**Comment by stefan-sherwood** [comment]
Made UI look more like AA page...

Local cam selection
image

Network checked
image

**Comment by rsjudka** [comment]
i like it! Give me some time to go through the code changes (curses github for screwing up the diff) and ill get it all merged up

**Comment by stefan-sherwood** [comment]
Thanks. It may be easier to use use `--word-diff` and diff that one file locally.

File: src/app/tabs/camera.cpp [hunk 3]
```diff
@@ -1,135 +1,361 @@
...
+            max_fit = resolution;
+        }
+    }
+    if (max_fit.isValid()) {
+        qDebug() << "Local cam auto resolution" << max_fit << "to fit in" << window_size;
+        this->local_cam_settings.setResolution(max_fit);
+    }
+    else
+        qDebug() << "No suitable resolutions found to fit in" << window_size;
+    //    this->local_cam_settings.setPixelFormat(QVideoFrame::Format_YV12);
```
Comment (stefan-sherwood) [inline src/app/tabs/camera.cpp:317]: This is for @robert5974 to uncomment on his machine until we understand why his machine is mismatching formats.

**Comment by stefan-sherwood** [comment]
PR migrated to openDsh. Closing.

**Pull request closed without merging**